manifests/init.pp

```diff
@@ -1,24 +1,52 @@
+# @summary Ensures a couple of mandatory files are present before managing their content.
 #
-# == Class: kmod
-#
-# Ensures a couple of mandatory files are present before managing their
-# content.
-#
+# @param owner Default owner for all files (set via Hiera to allow defaults on all defined types)
+# @param group Default group for all files (set via Hiera to allow defaults on all defined types)
+# @param directory_mode Default mode for all directories (set via Hiera to allow defaults on all defined types)
+# @param file_mode Default mode for all regular files (set via Hiera to allow defaults on all defined types)
+# @param exe_mode Default mode for all executable files (set via Hiera to allow defaults on all defined types)
+# @param list_of_aliases Hash of [`kmod::alias`](#kmodalias) resources
+# @param list_of_blacklists Hash of [`kmod::blacklist`](#kmodblacklist) resources
+# @param list_of_installs Hash of [`kmod::install`](#kmodinstall) resources
+# @param list_of_loads Hash of [`kmod::load`](#kmodload) resources
+# @param list_of_options Hash of [`kmod::option`](#kmodoption) resources
+# @param modprobe_d Location of `modprobe.d` directory
+# @param modprobe_d_files Default files to create in `modprobe.d` directory
 #
+# @example
+#   include kmod
 class kmod (
-  Hash $list_of_aliases    = {},
-  Hash $list_of_blacklists = {},
-  Hash $list_of_installs   = {},
-  Hash $list_of_loads      = {},
-  Hash $list_of_options    = {},
+  # Defaults for these are in Hiera to enable using them as
+  # defaults in the various defined types via `lookup()`.
```

Comment:
but why do you want to use lookup()? all the defined resources require the kmod class, you could reference the variables from that class.

Reply:
I was getting failures from rspec when I tried that.

Reply:
Testing again, if I do this:
```diff
diff --git a/manifests/load.pp b/manifests/load.pp
index 6823751..854c9e5 100644
--- a/manifests/load.pp
+++ b/manifests/load.pp
@@ -16,10 +16,10 @@ define kmod::load (
   Stdlib::Unixpath           $file     = '/etc/modules',
   # Defaults for these are set in Hiera so they can be shared with the `kmod` class.
   # lint:ignore:lookup_in_parameter
-  Optional[String[1]]        $owner    = lookup('kmod::owner', String[1], 'first', undef),
-  Optional[String[1]]        $group    = lookup('kmod::group', String[1], 'first', undef),
-  Optional[Stdlib::Filemode] $mode     = lookup('kmod::file_mode', String[1], 'first', undef),
-  Optional[Stdlib::Filemode] $exe_mode = lookup('kmod::exe_mode', String[1], 'first', undef),
+  Optional[String[1]]        $owner    = $kmod::owner,
+  Optional[String[1]]        $group    = $kmod::group,
+  Optional[Stdlib::Filemode] $mode     = $kmod::file_mode,
+  Optional[Stdlib::Filemode] $exe_mode = $kmod::exe_mode,
   # lint:endignore
 ) {
   include kmod
```
I get `error during compilation: Evaluation Error: Error while evaluating a Resource Statement, Evaluation Error: Unknown variable: 'kmod::owner'.`

Alternatively, I could drop the settings on the individual defined types and directly reference `$kmod::owner` (etc.) directly in the defined types, but then we lose the ability to override those settings for the individual types.

Reply:
@bastelfreak I removed the `lookup()` calls (and the ability to override permissions on individual resources).  Please take another look when you have a moment.